src/shared.rs

```diff
@@ -0,0 +1,211 @@
+use std::sync::{Arc, Mutex};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::ops::Deref;
+use std::sync::RwLock;
+use std::sync::mpsc::{channel, Receiver, Sender};
+use {Future, Poll, Async};
+use task::{self, Task};
+
+#[derive(Debug)]
+pub struct SharedItem<T> {
+    item: Arc<T>,
+}
+
+impl<T> Clone for SharedItem<T> {
+    fn clone(&self) -> Self {
+        SharedItem { item: self.item.clone() }
+    }
+}
+
+#[derive(Debug)]
+pub struct SharedError<E> {
+    error: Arc<E>,
+}
+
+impl<T> Clone for SharedError<T> {
+    fn clone(&self) -> Self {
+        SharedError { error: self.error.clone() }
+    }
+}
+
+impl<T> SharedItem<T> {
+    fn new(item: T) -> Self {
+        SharedItem { item: Arc::new(item) }
+    }
+}
+
+impl<E> SharedError<E> {
+    fn new(error: E) -> Self {
+        SharedError { error: Arc::new(error) }
+    }
+}
+
+struct SyncedInner<F>
+    where F: Future
+{
+    original_future: F,
+    result: Option<Result<Async<SharedItem<F::Item>>, SharedError<F::Error>>>,
+    tasks_receiver: Mutex<Receiver<Task>>,
+}
+
+struct Inner<F>
+    where F: Future
+{
+    synced_inner: RwLock<SyncedInner<F>>,
```

Comment:
This implementation is making some assumptions about `RwLock` and the various return values of `try_read` and `try_write`. That is, it assumes that if `Err` is returned that it's guaranteed that another task has the lock, as opposed to, for example, also being in contention for the same lock.

Put another way, I could imagine an implementation where two concurrent calls to `try_read` would have one fail and one succeed. 

I think a situation like that could cause this implementation to deadlock. Perhaps we could vendor a copy with known semantics around this form of contention to ensure we're getting what we want?